### PR TITLE
Docker compose -- package reorganization

### PR DESCRIPTION
**Pull request by glciampaglia**
Reorganized project / package structure to decouple source from data / models (to be loaded from cache automatically), improved docker container image building processes.

## Summary of changes
- Added top-level package `dante`, renamed project
- Created `dante.app` and `dante.osomerank` sub-packages
- In `dante.osomerank`:
  - Fixed all imports in `dante.osomerank` and reorganized imports section of each file (let's please keep these tidy!)
  - Moved data/model folders to user cache dir (determined automatically from XDG specs using [platformdirs](https://platformdirs.readthedocs.io/en/latest/api.html#unix-linux)
  - Do not load data/models on import, added `load_*_{data,models}()` function to ER/AD/TD models and `load_all()` functions as convenient way to load everything
    - Use `dante.osomerank.utils.fetchfroms3` across all modules for loading data / models from S3 bucket
  - Cleaned up model `.audience_diversity`, `.topic_diversity`, and `.elicited_response` modules under `dante.osomerank`. 
    - Use regular expressions in `ad_predictions` (much faster than list scan)
  - Updated logging format, use `dante.osomerank.utils.get_logger(__name__)` to identify which module is logging what
  - Use log rotation (keep 100MB worth of logs) and store logs in user log dir
  - Remove `config.ini` from version control, added sample config file with data/models path (but no S3 / DB secrets)
  - Search multiple locations for configuration files (in the following order):
     - Current working directory
     - User config dir (from XDG specs)
     - Site config dir (ditto)
     - Package location
  - If no config file is present, the pre-filled sample config file is automatically copied over to user config dir the first time the package is imported.
- In `dante.app`:
  - Load model artifacts in scorer workers only (using `dante.osomerank.load_all()`, see above)
  - Fixed imports and general code cleanup
  - Fixed typo REDIS_URL &rarr; REDIS_CONNECTION_STRING in ranking server
  - Added fakeredis dependency
  - Make embedded URLs JSON encodable for celery
  - Use ThreadPoolExecutor to submit requests to scorer workers from ranker
- Docker builds
  - Use multi-stage builds for docker containers (reduced image size from 12GB to 5.4GB)
  - Mount configuration file using bind mount
  - Use common image for all ranker and scorer worker   (see `Dockerfile` at the root of the repo)
 
## Breaking changes
To make sure everything works correctly after merging, make sure to go over the following steps:
- The project name has changed so you will need to run `poetry install` again to recreate a new virtualenv for the depenencies.
- Move `dante/osomerank/data/` and `dante/osomerank/models/` to your user cache dir (see on the platformdirs doc for where to find it on [mac os](https://platformdirs.readthedocs.io/en/latest/api.html#platformdirs.macos.MacOS.user_cache_dir), [windows](https://platformdirs.readthedocs.io/en/latest/api.html#platformdirs.windows.Windows.site_cache_dir), and [linux](https://platformdirs.readthedocs.io/en/latest/api.html#platformdirs.unix.Unix.user_cache_dir)). There is no `$version`, while the `$appname` is `dante`. So for example on my Linux laptop the user cache dir is located at `/home/giovanni/.cache/dante/`.
- If you don't move data and models to the cache, run this to automatically fetch files from the S3 bucket (requires `S3` configuration section in your config.ini):
```python
import dante.osomerank
dante.osomerank.load_all()
```

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -33,38 +35,59 @@ services:
...
       REDIS_CONNECTION_STRING: redis://redis:6379
       CELERY_BROKER: redis://redis-celery-broker:6380
       CELERY_BACKEND: redis://redis-celery-broker:6380
-    command: ["poetry", "run", "uvicorn", "--host", "0.0.0.0", "--port", "5001", "ranking_server:app"]
+    command: ["uvicorn", "--host", "0.0.0.0", "--port", "5001", "dante.app.ranking_server.ranking_server:app"]
     ports:
       - "5001:5001"
+    volumes:
+      - type: bind
+        source: ${PWD}/config.ini
```
Comment (btrantruong) [inline docker-compose.yml:50]: There seems to be a missing config.ini file? When I run `make run` it looks for an config.ini in the root project folder. 
---
Error response from daemon: invalid mount config for type "bind": bind source path does not exist: /host_mnt/Users/baott/osome-csdl-ranking-challenge/config.ini

File: dante/osomerank/sample_posts/sample_post_facebook.json
Comment (btrantruong) [inline dante/osomerank/sample_posts/sample_post_facebook.json:1]: It's safe to delete files in this folder if you don't use them

**Review by btrantruong — APPROVED** [review]
I just have a question on the config file to troubleshoot docker compose but everything looks good. Thanks for the great work!

File: docker-compose.yml (hunk 1 above)
Comment (glciampaglia) [reply docker-compose.yml:50]: Yes you will need to supply a config.ini in the same folder where you run docker-compose, in this case the root of the repository. This is meant to supply the configuration without needing to rebuild the image, which will be helpful when deploying in production. But for now the contents of this file are just the same as the sample config file in the dante/osomerank. So you can just copy that there at the root of the repository. 

In the docker-compose file there is also another bind mount for the cache. This is also to keep the image size small and avoiding to rebuild images every time the cache is cleared. For that you will need to copy your `data/` and `models/` folder to your user cache dir as described in the "Breaking Changes" section above. Alternatively you can run `dante.osomerank.load_all()` once look into the logs the cache location there and then point the docker-compose file there. Just make sure to include your values in the `S3` section of the config file.

File: dante/osomerank/sample_posts/sample_post_facebook.json
Comment (glciampaglia) [reply dante/osomerank/sample_posts/sample_post_facebook.json:1]: OK will do!

**Pull request merged**